### ISSUE TITLE
test_find_program() fails if .py isn't in PATHEXT

### ISSUE THREAD

**Issue by jon-turney** [mesonbuild/meson#4355]
`test_find_program()` (a windows-only test) fails if `.py` isn't in `PATHEXT`

```
C:\meson>py run_unittests.py WindowsTests.test_find_program
F
Stdout:
Program cmd found: YES (C:\WINDOWS\system32\cmd.EXE)
Program cmd.exe found: YES (C:\WINDOWS\system32\cmd.exe)
Program C:\WINDOWS\system32\cmd found: YES (C:\WINDOWS\system32\cmd.exe)
Program C:\cygwin64\wip\meson\test cases/windows\8 find program\test-script found: YES (python C:\cygwin64\wip\meson\test cases/windows\8 find program\test-script)
Program C:\cygwin64\wip\meson\test cases/windows\8 find program\test-script-ext.py found: YES (C:\Users\jon\AppData\Local\Programs\Python\Python36\python.exe C:\cygwin64\wip\meson\test cases/windows\8 find program\test-script-ext.py)
Program test-script-ext found: NO

======================================================================
FAIL: test_find_program (__main__.WindowsTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "run_unittests.py", line 3041, in test_find_program
    self.assertTrue(prog.found(), msg='test-script-ext not found in PATH')
AssertionError: False is not true : test-script-ext not found in PATH

Stdout:
Program cmd found: YES (C:\WINDOWS\system32\cmd.EXE)
Program cmd.exe found: YES (C:\WINDOWS\system32\cmd.exe)
Program C:\WINDOWS\system32\cmd found: YES (C:\WINDOWS\system32\cmd.exe)
Program C:\cygwin64\wip\meson\test cases/windows\8 find program\test-script found: YES (python C:\cygwin64\wip\meson\test cases/windows\8 find program\test-script)
Program C:\cygwin64\wip\meson\test cases/windows\8 find program\test-script-ext.py found: YES (C:\Users\jon\AppData\Local\Programs\Python\Python36\python.exe C:\cygwin64\wip\meson\test cases/windows\8 find program\test-script-ext.py)
Program test-script-ext found: NO

----------------------------------------------------------------------
Ran 1 test in 0.059s

FAILED (failures=1)

C:\meson>echo %PATHEXT%
.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC

C:\meson>set PATHEXT=%PATHEXT%;.py

C:\meson>py run_unittests.py WindowsTests.test_find_program
.
----------------------------------------------------------------------
Ran 1 test in 0.025s

OK
```

**Comment by jon-turney** [comment]
This test case seems a little odd.  `find_program('foo')` will (sometimes) find `foo.py` on windows, but not on any other platform?

**Comment by nirbheek** [comment]
Yes, that's how Windows `PATH` searching works. There's no executable bit, so executables are identified by their extension. The default value for `PATHEXT` is `.COM;.EXE;.BAT;.CMD` and if you want python scripts to be identified as executables, you have to append `;.PY` to that.

Of course this is optional, and the reason why the test contains this is just to ensure that this edge-case works (I'm sure some windows-only meson projects will want to assume that `find_program('foo')` can find `foo.py` in the correct configuration). 

Nothing in Meson depends on it, since we expect people to look for scripts with `find_program('foo.py')` since that is the cross-platform way of doing it. We can just skip that line of the test if `.PY` is not in `PATHEXT`, but we should ensure that at least one of our CI jobs has `.PY` in `PATHEXT`.

**Comment by nirbheek** [comment]
Also, Meson supports looking for scripts with a `#!.* python` named `foo` with `find_program('foo')` through manual listing of directories specified in `PATH` because a lot of Linux-origin projects tend to install scripts without extensions on all platforms (f.ex., glib).

**Comment by NNemec** [comment]
As it seems, this issue reappeared. Looking at the build pipeline, This commit did, in fact, address the issue, but then, a few commit later, the same error messages reappeared on the master branch without an accidental revert or a similar obvious reason.

**Comment by nirbheek** [comment]
Reopening as per https://github.com/mesonbuild/meson/pull/5564#issuecomment-506522689.

**Comment by brunolopesdsilv** [comment]
Sill reproducible in meson 1.4. meson can't find yelp-build(.py) provided by yelp-tools MSYS2 package.

**Comment by jon-turney** [comment]
I apologise again for the terrible title I chose initially for this issue.

It seems like find_program() should never use PATHEXT on Windows, as that leads to find_program("something") working, when a well formed meson.build should say find_program("something.py"), but making that change would break existing meson.build which happen to work in some places...

**Comment by eli-schwartz** [comment]
Should people have to specify `find_program('foo.exe') on Windows and `find_program('foo')` on Linux?

**Comment by eli-schwartz** [comment]
> Still reproducible in meson 1.4. meson can't find yelp-build(.py) provided by yelp-tools MSYS2 package.

The msys2 package list at packages.msys2.org lists that it provides a file called usr/bin/yelp-build with no extension, not even a .py extension. Is this correct?

**Comment by eli-schwartz** [comment]
In that case I'm not sure I understand the problem. At least this isn't related to PATHEXT.

**Comment by jon-turney** [comment]
> Should people have to specify `find_program('foo.exe') on Windows and `find_program('foo')` on Linux?

You're quite right. Pehaps I mean "find_program() shouldn't use a unspecified, non-default value of PATHEXT inherited from the environment"?

**Comment by bonzini** [comment]
>  "find_program() shouldn't use an unspecified, non-default value of PATHEXT inherited from the environment"?

What Meson does is pass the value returned by `shutil.which` to `_search_windows_special_cases` which does two things:
* check for extensions in `self.windows_exts`
* tries to apply the shebang if it's not one of those extensions.

Not using PATHEXT would probably fix some things and break others, so I'm hesitant to change that. I would like to understand what is a case in which the above "trick" breaks.

**Comment by jon-turney** [comment]
I think my concern was that someone could write a meson.build which contains `find_program('some-script")` where the file is named `some-script.py`, which would happen to work for them because they happen to have that extension in their PATHEXT, but fail elsewhere.

It might well be that I was mistaken, or that things have changed over the intervening ~8 years that this is no longer an issue.

> Not using PATHEXT would probably fix some things and break others, so I'm hesitant to change that.

I guess as an interim step we could apply `shutil.which` with default and inherited environment PATHEXT values and warn if it gives different answers.
